File: iosInblo/iosInblo/App/messages/MessagesViewModel.swift
//
//  MessagesViewModel.swift
//  iosInblo
//
//  Created by Jhondee Diaz on 2/9/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import shared
import SwiftUI

class MessagesViewModel: ObservableObject {
    
    @AppStorage("stableId") var stableId: Int = 0
    @AppStorage("userId") var userId: Int = 0
    
    let fetchMessages: GetAllMessages
    let sendMessage: SendMessageToUser
    let removeMessage: RemoveMessage
    let getUsernames: GetUsernames
    let searchHorses: SearchHorses
    
    @Published var alert: AlertContent?

    @Published var messages = [Message]()
    @Published var users    = [UserDto]()
    @Published var horses   = [Horse]()
    
    init(
        fetchMessages: GetAllMessages,
        sendMessage: SendMessageToUser,
        removeMessage: RemoveMessage,
        getUsernames: GetUsernames,
        searchHorses: SearchHorses
    ){
        self.fetchMessages = fetchMessages
        self.sendMessage = sendMessage
        self.removeMessage = removeMessage
        self.getUsernames = getUsernames
        self.searchHorses = searchHorses
        
        self.loadUserNames(stabId: String(stableId), exId: String(userId))
        self.loadHorses()
        
    }
    
    
    func loadMessages(userId: Int){
        do {
            fetchMessages.invoke(userId: Int32(userId)).collectCommon(coroutineScope: nil, callback: { dataState in
                if(dataState is DataStateData){
                    let messages = dataState?.value(forKey: "data") as! [Message]
                    self.messages = messages
                    print(messages)
                }
            })
        }
    }
    
    func loadHorses(
        isArchived: Bool = false
    ){
        do {
            searchHorses.invoke(
                offset: 0,
                query: nil,
                isArchived: isArchived,
                stableId: Int32(stableId)
            ).collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                if dataState != nil {
                    
                    if(dataState is DataStateData){
                        
                        let horses = dataState?.value(forKey: "data") as! [Horse]
                        self.horses = horses
                    }
                
                }
            })
        }
    }
    
    func loadUserNames(
        stabId: String? = nil,
        exId: String? = nil
    ){
        do {
            getUsernames.invoke(stableId: stabId, excludeId: exId).collectCommon(coroutineScope: nil, callback: { dataState in
                if dataState != nil {
                    if(dataState is DataStateData){
                        print(dataState)
                        let getUserResponse = dataState?.value(forKey: "data") as! GetUsersResponse
                        self.users = getUserResponse.data
                    }
                }
                
            })
        }
    }
    
    func removeMessage(
        messageId: Int,
        userId: Int
    ){
        do{
            removeMessage.invoke(messageId: Int32(messageId)).collectCommon(coroutineScope: nil, callback: { dataState in
                if dataState != nil {
                    if(dataState is DataStateData){
                        print(dataState)
                        let response = dataState?.value(forKey: "data") as! BooleanResponse
                        print(response)

                        self.alert = AlertContent(title: "", body: "message deleted successfully!", type: getAlertTypeByCode(200))
                        self.loadMessages(userId: userId)

                    }
                }
                
            })
        }
    }
    
    func addMessage(
        userId: Int,
        stableId: Int,
        senderId: Int,
        recipientId: String? = nil,
        horseId: String? = nil,
        horseName: String,
        notificationType: String,
        title: String,
        memo: String,
        isRead: String,
        messageId: String? = nil
    ){
        do {
            sendMessage.invoke(stableId: Int32(stableId), senderId: Int32(senderId), recipientId: recipientId, horseId: horseId, horseName: horseName, notificationType: notificationType, title: title, memo: memo, isRead: isRead, messageId: messageId).collectCommon(coroutineScope: nil, callback: { dataState in
                    if(dataState is DataStateData){
                        self.alert = AlertContent(title: "Add Message", body: "Message Added Successfully!", type: getAlertTypeByCode(200))
                        print(dataState)
                        let message = dataState?.value(forKey: "data") as! Message
                        
                        self.loadMessages(userId: userId)
                        print("alert is -> \(self.alert)")
                    }
            })
        }
    }
}

File: androidInblo/src/main/java/com/colinjp/inblo/android/presentation/ui/messages/MessagesViewModel.kt
package com.colinjp.inblo.android.presentation.ui.messages

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.colinjp.inblo.android.presentation.ui.horse_list.SortOrder
import com.colinjp.inblo.domain.model.Horse
import com.colinjp.inblo.domain.model.Message
import com.colinjp.inblo.domain.model.User
import com.colinjp.inblo.interactors.horse_list.GetUsernames
import com.colinjp.inblo.interactors.horse_list.SearchHorses
import com.colinjp.inblo.interactors.message.GetAllMessages
import com.colinjp.inblo.interactors.message.MarkMessageAsRead
import com.colinjp.inblo.interactors.message.RemoveMessage
import com.colinjp.inblo.interactors.message.SendMessageToUser
import com.colinjp.inblo.util.DataState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

@HiltViewModel
class MessagesViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle,
    private val getAllMessages: GetAllMessages,
    private val sendMessageToUser: SendMessageToUser,
    private val markMessageAsRead: MarkMessageAsRead,
    private val removeMessage: RemoveMessage,
    private val getUsernames: GetUsernames,
    private val searchHorses: SearchHorses
): ViewModel() {

    val messages = MutableStateFlow<DataState<List<Message>>>(DataState.Empty)

    val horses = MutableStateFlow<DataState<List<Horse>>>(DataState.Empty)
    val users = MutableStateFlow<DataState<List<User>>>(DataState.Empty)

    fun getHorses(sortOrder: SortOrder, stableId: Int){
        viewModelScope.launch(IO){
            searchHorses.invoke(stableId = stableId
            ).onEach {
                horses.value = it
            }.launchIn(viewModelScope)
        }
    }

    fun getUsernames(stableId: String? = null,excludeId: String? = null){
        viewModelScope.launch(IO) {
            getUsernames.invoke(stableId,excludeId).onEach { userResponse ->
                when(userResponse){
                    is DataState.Data -> {
                        users.value = DataState.Data(userResponse.data.data.map { userDto ->  User.createFromDto(userDto) })
                    }
                    DataState.Empty -> {

                    }
                    is DataState.Error -> {
                        //TODO:

                    }
                    is DataState.Loading -> {
                        //TODO:

                    }
                }
            }.launchIn(viewModelScope)
        }
    }

    private val eventChannel = Channel<MessageEvent>()
    val eventFlow = eventChannel.receiveAsFlow()

    fun getMessages(userId: Int){
        viewModelScope.launch(IO){
            getAllMessages.invoke(userId).onEach {
                messages.value = it
            }.launchIn(viewModelScope)
        }
    }

    fun sendMessage(
        stableId: Int,
        senderId: Int,
        recipientId: Int?,
        horseId: Int?,
        horseName: String,
        notificationType: String,
        title: String,
        memo: String,
        isRead: String,
        messageId: Int?
    ){
        viewModelScope.launch(IO){
            sendMessageToUser.invoke(stableId,senderId, recipientId.toString(), horseId.toString(), horseName, notificationType, title, memo, isRead,messageId.toString())
                .onEach {
                    when(it){
                        is DataState.Data -> {
                            Timber.v(it.toString())

                            if(messageId == null ) eventChannel.send(MessageEvent.Success("Message successfully sent!")) else eventChannel.send(MessageEvent.Success("Message successfully updated!"))
                        }
                        DataState.Empty -> {

                        }
                        is DataState.Error -> {
                            //TODO:

                            Timber.e(it.toString())
                        }
                        is DataState.Loading -> {
                            //TODO:

                        }
                    }
                }.launchIn(viewModelScope)
        }
    }

    fun markMessageAsRead(
        messageId: Int
    ) {
        viewModelScope.launch(IO){
            markMessageAsRead.invoke(messageId).onEach {
                when(it){
                    is DataState.Data -> {
                        Timber.v("Message $messageId marked as read")
                        messages.value = it
                    }
                    DataState.Empty -> {

                    }
                    is DataState.Error -> {
                        //TODO:

                        Timber.e(it.exception)
                    }
                    is DataState.Loading -> {
                        //TODO:

                    }
                }
            }
        }
    }

    fun deleteMessage(
        messageId: Int
    ){
        viewModelScope.launch(IO){
            removeMessage.invoke(messageId).onEach {
                when(it){
                    is DataState.Data -> {
                        Timber.v("Message $messageId removed")
                        eventChannel.send(MessageEvent.Success("Message Successfully removed!"))
                    }
                    DataState.Empty -> {

                    }
                    is DataState.Error -> {
                        //TODO:

                        Timber.e(it.exception)
                    }
                    is DataState.Loading -> {
                        //TODO:

                    }
                }
            }.launchIn(viewModelScope)
        }
    }

    sealed class MessageEvent {
        data class Success(val message: String): MessageEvent()
        data class Error(val errorMessage: String): MessageEvent()
        data class Loading(val isLoading: Boolean): MessageEvent()
    }
}